README.md
# miner_watchdog
A client part of application for monitoring cryptocurrency miners written in Python. At the moment, only Claymore miners are supported, since I'm using it's API for communication with the miner app.

Visit [project page](https://malja.github.io/miner_watchdog/) for more documentation.


Changelog
0.0.1
=====

- First version with basic functionality. Client uploads miner data (hashrate, runtime, name, version, ...) to server
  for monitoring purposes.

0.0.2
=====

- Changed project structure - almost everything is in client folder.
- Client uploads "interval" from its config. 
- Added App class with callbacks from twisted's Factory class

client/ping.py
import platform
import subprocess

"""
Currently unused code. It was used to check connection to router/website in case of unavailability of main server.
"""

def ping(host, network_timeout=3):
    """
    Send a ping packet to the specified host, using the system "ping" command.
    Source from: https://stackoverflow.com/a/48767169/2187426
    :param str host: IP address or hostname (website address) as ping target.
    :param int network_timeout: Number of seconds to wait before timeout.
    :returns: True if target responded before timeout, False otherwise.
    """
    
    args = [
        'ping'
    ]

    platform_os = platform.system()

    if platform_os == 'Windows':
        args.extend(['-n', '1'])
        args.extend(['-w', str(network_timeout * 1000)])
    elif platform_os in ('Linux', 'Darwin'):
        args.extend(['-c', '1'])
        args.extend(['-W', str(network_timeout)])
    else:
        raise NotImplementedError('Unsupported OS: {}'.format(platform_os))

    args.append(host)

    try:
        if platform_os == 'Windows':
            output = subprocess.run(args, check=True, universal_newlines=True).stdout

            if output and 'TTL' not in output:
                return False
        else:
            subprocess.run(args, check=True)

        return True
    
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired):
        return False

client/config.py
import configparser
from os.path import exists, isfile
from collections import namedtuple


class Config(object):
    """
    Class for loading INI configuration file into a namedtuple.
    """

    def __init__(self, logger):

        self._path = ""
        self._content = None
        self._parser = configparser.ConfigParser()
        self.logger = logger.getChild("config")

    def open(self, path):

        if not (exists(path) and isfile(path)):
            self.logger.error("Configuration file '{}' does not exist.".format(path))
            return False

        self.path = path

        if len(self._parser.read(path, "UTF8")) != 1:
            self.logger.error("Unable to read configuration file '{}'".format(path))
            return False

        data = {}

        for section in self._parser.sections():
            data[section] = {}

            for key in self._parser.options(section):
                data[section][key] = self._parser.get(section, key)

        self._content = self._convert_dict_to_namedtuple(data)

    def _convert_dict_to_namedtuple(self, d):
        """
        
        """
        for k, v in d.items():
            if isinstance(v, dict):
                d[k] = self._convert_dict_to_namedtuple(v)
        return namedtuple('object', d.keys())(*d.values())

    def __getattr__(self, item):
        if hasattr(self._content, item):
            return getattr(self._content, item)

        raise AttributeError()

client/__init__.py
import os
import time

from twisted.internet import reactor
from twisted.internet.error import ConnectionDone

from client.config import Config
from client.logs import SetupLogs
from client.connection import ClaymoreConnectionFactory, ServerReport


class App(object):
    """
    This class represents the client. It is called when application starts and performs all connections, logging etc.
    """

    """
    Current client version. It is checked with server app to redirect it to right API version.
    """
    version = "0.0.2"

    def __init__(self, path):
        """
        Create client.
        :param str path: Path to client.py file. It is used as a base for all relative paths.
        """
        self.path = path

        self.logger = SetupLogs(self.path)

        # Load configuration
        self.config = Config(self.logger)
        self.config.open(os.path.abspath(os.path.join(self.path, "./config.ini")))

    def run(self):
        """
        Starts a client.
        """
        self.logger.info("Started up after boot up.")
        
        # Create inner TCP connection factory
        reactor.connectTCP(self.config.client.host, int(self.config.client.port), ClaymoreConnectionFactory(self))
        reactor.run()

    def onConnectionClosed(self, reason, connector):
        """
        Callback executed every time connection with claymore miner is closed.
        :param object reason: One of twisted.internet.error classes with reason for closing connection.
        :param object connector: Object for reconnecting.
        """

        # Check closing reason
        if isinstance(reason.type(), ConnectionDone):

            # Everything went ok. 
            self.logger.info("Waiting for {} s".format(self.config.client.interval))
            time.sleep(int(self.config.client.interval))
            
            # Reconect
            connector.connect()

        else:
            
            # Some kind of error
            self.logger.error("Connection lost: {}".format(reason))

    def onConnectionFailed(self, reason, connector):
        """
        Callback called when connection to claymore miner fails.
        :param object reason: One of twisted.internet.error reasons.
        :param object connector: Object for reconnecting.
        """
        pass

    def onMinerResponse(self, status):
        """
        Callback called when claymore miner responses to a message.
        :param client.connection.ClaymoreResponse status: Response data.
        """

        if not status:
            self.logger.error("Status error: {}".format(status.getError()))
            return False

        self.logger.info("Hashrate: {} H/s".format(status.hashrate()))

        # Restart Miner
        if status.hashrate() < int(self.config.client.hashrate):
            self.logger.info("Hashrate is too low. Restarting computer.")
            os.system("r")

        # Send data to monitoring server
        report = ServerReport(status, self)
        report.send()

        self.logger.debug("Report sent to collecting server.")

        if report.return_code() != 200:
            self.logger.error("Server responded with error: Code: {} - {}".format(report.return_code(), report.response()))
            return False

        else:
            self.logger.info("Server responded with success code.")
            return True


client/connection.py
import json
import requests
import os
import time

from twisted.internet.protocol import Protocol, ClientFactory


class ServerReport(object):
    """
    Object used as wrapping around HTTP/S connection to monitoring server.
    """

    def __init__(self, status, app):
        """
        :param client.connection.ClaymoreResponse status: Class with response from claymore miner.
        :param client.App app: App instance.
        """

        self.data = status
        self.app = app

        self._status_code = 0
        self._response = None

    def send(self):
        """
        Send all data to server via POST.
        """

        payload = {
            "name": self.app.config.client.name,
            "version": self.app.version,
            "secret": self.app.config.client.secret,
            "interval": self.app.config.client.interval,
            "data": json.dumps({
                "version": self.data.version(),
                "runtime": self.data.runtime(),
                "hashrate": self.data.hashrate(),
                "pool": self.data.pool(),
                "gpus": self.data.gpuData()
            })
        }

        try:
            r = requests.post(self.app.config.server.address, data=payload)
            self._response = r.text
            self._status_code = r.status_code

        except (requests.exceptions.ConnectionError, requests.exceptions.ConnectTimeout) as e:
            self.app.logger.error("Server connection failed {}. Restarting computer".format(e))
            os.system("r")

        except Exception as e:
            self.app.logger.error("Unhandled connection error: {}".format(e))

    def __bool__(self):
        return self._status_code == 200

    def response(self):
        """
        Text printed out by server. Mainly for debugging purpuses (in default, server does not output anything).
        """
        return self._response

    def return_code(self):
        return self._status_code


class ClaymoreRequest(object):
    """
    Class for hiding all code for creating a JSON string with a message sent to Claymore miner. All methods are pure
    static.
    """

    @staticmethod
    def _message(method, params = None):
        """
        This method build a JSON string from method and additional parameters. You should not call this method directly
        but rather use one of "one purpose" one listed below.
        :param str method: One of supported methods. It sets type of message for claymore.
        :param list params: Array of optional parameters required for selected method.
        :return: JSON string.
        """

        message = {
            "id": 0,
            "jsonrpc": "2.0",
            "method": method,
        }

        if params:
            message["params"] = params

        return json.dumps(message, ensure_ascii=True).encode("ascii")

    @staticmethod
    def status():
        """
        Requests a recent miner stats - hashrate, runtime etc.
        :return: JSON string.
        """

        method = "miner_getstat1"
        return ClaymoreRequest._message(method)

    @staticmethod
    def restart():
        """
        Restarts Claymore miner.
        :return: JSON string.
        """

        method = "miner_restart"
        return ClaymoreRequest._message(method)

    @staticmethod
    def reboot():
        """
        Calls reboot.bat (Windows) or reboot.sh (Linux).
        :return: JSON string.
        """

        method = "miner_reboot"
        return ClaymoreRequest._message(method)

    @staticmethod
    def gpu(index, state):
        """
        Controls GPU - disable, ETH-only and dual mining mode.
        :param int index: Index of the card as seen by the miner. Or -1 for all cards.
        :param int state: 0 (disabled), 1 (ETH only), 2 (dual mining)
        :return: JSON string.
        """

        if index < -1:
            raise ValueError("Index have to be bigger than -1.")

        if state not in [0, 1, 2]:
            raise ValueError("State is required to be one of 0,1 or 2")

        method = "control_gpu"
        params = [index, state]
        return ClaymoreRequest._message(method, params)


class ClaymoreResponse(object):
    """
    Class wrapper around JSON response from claymore miner.
    """

    def __init__(self, data):
        """
        :param str data: RAW response as JSON string.
        """

        self._error = None
        self._version = None
        self._runtime = 0
        self._hashrate = 0
        self._gpu = []
        self._pool = None

        self._parse(data)

    def _parse(self, data):
        """
        Do not call this method directly.
        """

        # Claymore returns data as JSON string
        data_dict = json.loads(data.decode("utf-8"))

        # Get error message
        self._error = data_dict["error"]

        # Check error
        if self._error:
            return False

        result = data_dict["result"]

        # Get miner version
        self._version = result[0]

        # Get runtime
        self._runtime = int(result[1])

        # Get hashrate. It is in array with valid and invalid shares.
        # Hashrate is in kH/s
        # hashrate;valid shares;invalid shares
        shares = result[2].split(";")
        self._hashrate = int(shares[0])*1000

        # Hashrates are in kH/s
        gpu_hashrate = result[3].split(";")

        # Result[6] is list of temps and fanspeeds divided by ;
        # temp1;fan1;temp2;fan2 ...
        gpu_temp_fan = result[6].split(";")

        for index, hashrate in enumerate(gpu_hashrate):
            self._gpu.append(
                {
                    "index": index,
                    "hashrate": int(hashrate)*1000,
                    "temperature": int(gpu_temp_fan[index]),
                    "fanspeed": int(gpu_temp_fan[index+1])
                }
            )

        # Get Pool address
        self._pool = result[7]

        return True

    def hasError(self):
        return self._error is not None

    def getError(self):
        return self._error

    def hashrate(self):
        return self._hashrate

    def pool(self):
        return self._pool

    def version(self):
        return self._version

    def runtime(self):
        return self._runtime

    def gpuData(self):
        """
        Get whole dictionary with GPU related data.
        :returns: List of GPU dicts.
        """
        return self._gpu

    def gpu(self, index):
        """
        Returns data about GPU with selected index.
        :param int index: GPU number (as seen by miner).
        :returns: GPU dict.
        """
        return self._gpu[index]

    def __bool__(self):
        return self._error is None


class ClaymoreConnection(Protocol):

    def sendMessage(self, message):
        self.app.logger.debug("Sending message to Claymore miner.")
        self.transport.write(message)


    def dataReceived(self, data):

        self.app.logger.debug("Data received from Claymore miner.")
        status = ClaymoreResponse(data)
        self.app.onMinerResponse(status)

    def connectionMade(self):
        self.app.logger.info("Connected to Claymore miner")
        self.sendMessage(ClaymoreRequest.status())


class ClaymoreConnectionFactory(ClientFactory):
    def __init__(self, app):
        self.app = app

        # Sleep for selected time to give miner time to start properly.
        time.sleep(int(self.app.config.client.interval))

    def buildProtocol(self, addr):
        protocol = ClaymoreConnection()
        protocol.app = self.app
        return protocol

    def clientConnectionLost(self, connector, reason):
        self.app.onConnectionClosed(reason, connector)

    def clientConnectionFailed(self, connector, reason):
        self.app.logger.error("Connection failed: ".format(reason))
        self.app.onConnectionFailed(reason, connector)
